src/uwtools/drivers/forecast.py

```diff
@@ -126,10 +131,64 @@ def resources(self) -> None:
         ???
         """
 
-    def run(self):
+    def run(
+        self,
+        config_file: str,
+        forecast_app: str,
+        forecast_model: str,
+        dry_run: bool = False,
+        log_file: Optional[str] = None,
+        machine: Optional[str] = None,
+        quiet: bool = False,
+        verbose: bool = False,
+    ) -> None:  # pragma: no cover
         """
         Runs FV3.
         """
+        # set up logging
+        name = f"{inspect.stack()[0][3]}"
+        log = cli_helpers.setup_logging(
+            log_file=log_file or "/dev/stdout",
+            log_name=name,
+            quiet=quiet,
+            verbose=verbose,
+        )
+        # read in the config file
+        # self.create_model_config(config_file=config_file)
+        experiment_config = config.YAMLConfig(config_file)
+        # prepare directories
+        run_directory = experiment_config["parameters"]["run_directory"]
+        self.create_directory_structure(run_directory)
+
+        static_files = experiment_config["locations"]["static"]
+        self.stage_files(run_directory, static_files, link_files=False)
+        cycledep_files = experiment_config["locations"]["cycledep"]
+        self.stage_files(run_directory, cycledep_files, link_files=True)
+
+        experiment_resources = {
+            key: experiment_config["resources"][key] for key in experiment_config["resources"]
+        }
+        batch_script = self.batch_script(experiment_resources)
+        args = str(experiment_config["run_cmd"]["kwargs"])
+        run_command = self.run_cmd(
+            args,
+            run_cmd=experiment_config["run_cmd"]["run_cmd"],
+            exec_name=experiment_config["run_cmd"]["exec_name"],
+        )
+
+        if dry_run:
+            # Apply switch to allow user to view the run command of config
+            # This will not run the job.
+            log.info(f"Configuration: {forecast_app} {forecast_model} {machine}")
+            log.info(f"Run command: {run_command} {batch_script}")
+            return
+        # run the job
+        subprocess.run(
+            f"{run_command} {batch_script}",
+            capture_output=True,
+            check=False,
+            shell=True,
+        )
```

Comment:
I think that, if `run_command` fails in the subshell, `subprocess.run()` will raise an exception that -- since we've got `capture_output=True` -- we'll want to catch and handle by iterating over the captured output and logging what was captured. Otherwise, `uw` will terminate, but the user will not see any complaints that the program being run might have made and will possibly be mystified.

Well, I think that would be the _better_ way to do it, but I think the options are:

1. Remove `capture_output=True`, which will result in the output being printed directly to the appropriate stream (`stdout` and/or `stderr`, depending on how well-behaved the program we're running is); and just let whatever exception is raised terminate `uw`.
2. The better option IMO: Assign the return value of `subprocess.run()` to a variable, then extract the captured output from that variable and log the output, which we might want to do even if there is no error -- if the program being run produces information, it might be good to log it for the user to see. Then execute that inside a `try` block, handle whatever exception is raised (can't remember, but the `subprocess` docs should say), and then raise a (new!) subclass of `UWError` -- maybe `UWRuntimeError`? -- for code higher up the call stack to handle.

I can also think of two ways to deal with the fact that programs being run in a subshell may or may not use both `stdout` and `stderr`:

1. Set `stderr=STDOUT` as described in the [subprocess](https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module) docs, which will merge `stderr` into `stdout` so that there will be only one set of output to iterate over and log, which will look more like what a user would see in a terminal, where `stdout` and `stderr` messages are (normally, barring stream redirection in the shell) naturally interspersed.
2. Log e.g. `Captured output:`, iterate over and log `stdout` messages, log `Captured errors:`, iterate over and log `stderr` messages. This might be more intelligible to the user, as they won't see error and non-error output interspersed.

It's anybody's guess as to whether any given program will use `stdout` and/or `stderr` "correctly". Many NWP applications ignore this, but I think we should be ready for anything.

It might be worth checking to see whether there are other `subprocess.run()` calls that could benefit from the same treatment and, if so, factor this more robust handling out to a helper method that could be called here and from those other sites.

Reply:
Ah thats very fair, and an extension of your other comment. In my local branch I've placed in the quick solution of removing `capture_output=True` and replacing it with `stderr=subprocess.STDOUT` so there's a fix in place while I sort out the failing tests, but I'm hoping I can slip in adding more to `UWError`!